File: app/ClassScribe.Windows/src/ClassScribe.Core/SystemOutputCaptureAuthorization.cs
using System;
using System.Collections.Generic;

namespace ClassScribe.Core;

/// Ephemeral proof that an explicit system-output consent action occurred for
/// one exact capture attempt. It is intentionally not serializable and cannot
/// be rebuilt from session metadata.
public sealed class SystemOutputCaptureAuthorization
{
    internal SessionAttemptID Attempt { get; }

    internal Guid Nonce { get; }

    private SystemOutputCaptureAuthorization(SessionAttemptID attempt, Guid nonce)
    {
        Attempt = attempt;
        Nonce = nonce;
    }

    internal static SystemOutputCaptureAuthorization Create(
        SessionAttemptID attempt,
        Guid nonce) => new(attempt, nonce);
}

/// In-memory authority for system-output consent. The only issuer is the
/// explicit consent action in the product control-plane (and its deterministic
/// tests). Nothing is persisted or shared across attempts.
public sealed class SystemOutputCaptureAuthorizationAuthority
{
    private readonly object sync = new();
    private readonly Dictionary<SessionAttemptID, Guid> issuedNonces = [];

    public bool Accepts(
        SystemOutputCaptureAuthorization? authorization,
        SessionAttemptID attempt)
    {
        if (authorization is null)
        {
            return false;
        }

        lock (sync)
        {
            return authorization.Attempt == attempt
                && issuedNonces.TryGetValue(attempt, out var nonce)
                && nonce == authorization.Nonce;
        }
    }

    public void Invalidate(SessionAttemptID attempt)
    {
        lock (sync)
        {
            issuedNonces.Remove(attempt);
        }
    }

    /// Issues one in-memory capability only after the caller has completed the
    /// explicit system-output consent action for this exact attempt. There is
    /// deliberately no Bool-based or persisted issuance path.
    internal SystemOutputCaptureAuthorization IssueAfterExplicitUserConsent(SessionAttemptID attempt)
    {
        var nonce = Guid.NewGuid();
        lock (sync)
        {
            issuedNonces[attempt] = nonce;
        }

        return SystemOutputCaptureAuthorization.Create(attempt, nonce);
    }
}

File: app/MeetingTranscriber/ClassScribeSources/SystemOutputCaptureAuthorization.swift
import Foundation

/// An ephemeral proof that an explicit system-output consent action happened
/// for exactly one capture attempt. It deliberately has no Codable surface and
/// cannot be reconstructed from session metadata.
final class SystemOutputCaptureAuthorization: @unchecked Sendable {
    fileprivate let attempt: SessionAttemptID
    fileprivate let nonce: UUID

    fileprivate init(attempt: SessionAttemptID, nonce: UUID) {
        self.attempt = attempt
        self.nonce = nonce
    }
}

/// In-memory authority for system-output consent. The only issuer is the
/// explicit consent action in the product control-plane (and its deterministic
/// tests). A new authority is created per capture controller and nothing here
/// is persisted or shared across attempts.
final class SystemOutputCaptureAuthorizationAuthority: @unchecked Sendable {
    private let lock = NSLock()
    private var issuedNonces: [SessionAttemptID: UUID] = [:]

    func accepts(
        _ authorization: SystemOutputCaptureAuthorization?,
        for attempt: SessionAttemptID,
    ) -> Bool {
        guard let authorization else { return false }
        lock.lock()
        defer { lock.unlock() }
        return authorization.attempt == attempt
            && issuedNonces[attempt] == authorization.nonce
    }

    func invalidate(_ attempt: SessionAttemptID) {
        lock.lock()
        issuedNonces.removeValue(forKey: attempt)
        lock.unlock()
    }

    /// Issues one in-memory capability only after the caller has completed the
    /// explicit system-output consent action for this exact attempt. There is
    /// intentionally no Bool-based or persisted issuance path.
    func issueAfterExplicitUserConsent(for attempt: SessionAttemptID) -> SystemOutputCaptureAuthorization {
        let nonce = UUID()
        lock.lock()
        issuedNonces[attempt] = nonce
        lock.unlock()
        return SystemOutputCaptureAuthorization(attempt: attempt, nonce: nonce)
    }
}
